Inputting/Inputs/Input.cs

```diff
@@ -1,69 +1,75 @@
     {
         /// <summary>
         /// The input's representational display text.
         /// </summary>
         public string DisplayedText { get; }
         /// <summary>
-        /// The text describing the actual input instruction that should be executed.
+        /// The name of the input/button being triggered.
+        /// Conceptually, a touchscreen is also one big button.
         /// </summary>
-        public string EffectiveText { get; }
+        public string ButtonName { get; }
         /// <summary>
         /// The original text this input was parsed from.
         /// </summary>
         public string OriginalText { get; }
-        /// <summary>
-        /// Any additional data this input might have,
-        /// e.g. coords for touchscreen inputs or input intensity for analog sticks.
-        /// </summary>
-        public object AdditionalData { get; }
 
-        public Input(string displayedText, string effectiveText, string originalText, object additionalData)
+        public Input(string displayedText, string buttonName, string originalText)
         {
             DisplayedText = displayedText;
-            EffectiveText = effectiveText;
+            ButtonName = buttonName;
             OriginalText = originalText;
-            AdditionalData = additionalData;
         }
 
-        public bool Equals(Input other)
+        public override string ToString() => $"{DisplayedText}({ButtonName})";
+
+        #region polymorphic equals boilerplate
+
+        public override bool Equals(object? obj)
+        {
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            return obj.GetType() == GetType() && Equals((Input) obj);
+        }
+
+        public override int GetHashCode() => HashCode.Combine(DisplayedText, ButtonName, OriginalText);
+
+        private bool Equals(Input other)
         {
-            return DisplayedText.Equals(other.DisplayedText)
-                   && EffectiveText.Equals(other.EffectiveText)
-                   && OriginalText.Equals(other.OriginalText)
-                   && AdditionalData.Equals(other.AdditionalData);
+            return DisplayedText == other.DisplayedText
+                   && ButtonName == other.ButtonName
+                   && OriginalText == other.OriginalText;
         }
 
         /// <summary>
         /// Determines whether this input is effectively equal to another input,
         /// meaning if the inputs would cause the same action.
         /// This is done by only comparing the functional parts of this input.
         /// </summary>
-        /// <param name="other">input to check for effective equality</param>
+        /// <param name="obj">input to check for effective equality</param>
         /// <returns>whether the supplied input is effectively equal</returns>
-        public bool EqualsEffectively(Input other)
+        public virtual bool EqualsEffectively(Input? obj)
```

Comment:
Suggested name: IsSameButton.

Reply:
sounds a bit misleading. given that name, one could think there wouldn't be a difference between `a.IsSameButton(b)` and `a.ButtonName == b.ButtonName`

Reply:
"Effectively Equal" is very ambiguous. Is the point of this function to determine collisions? Because if so, a TouchscreenDrag and a Touchscreen input should collide even if they're not the same type

Reply:
it's meant for groupability. e.g. in democracy

Reply:
IsSameButtonGroup then?